Framework/Source/Graphics/Program/ProgramReflection.cpp

```diff
@@ -0,0 +1,1190 @@
+
+        case SLANG_RESOURCE_ACCESS_READ_WRITE:
+        case SLANG_RESOURCE_ACCESS_RASTER_ORDERED:
+            return ReflectionResourceType::StructuredType::Counter;
+        case SLANG_RESOURCE_ACCESS_APPEND:
+            return ReflectionResourceType::StructuredType::Append;
+        case SLANG_RESOURCE_ACCESS_CONSUME:
+            return ReflectionResourceType::StructuredType::Consume;
+        }
+    };
+
+    ReflectionVar::SharedPtr reflectVariable(VariableLayoutReflection* pSlangLayout, const ReflectionPath* pPath);
+    ReflectionType::SharedPtr reflectType(TypeLayoutReflection* pSlangType, const ReflectionPath* pPath);
+
+    // Once we've found the path from the root down to a particular leaf
+    // variable, `getDescOffset` can be used to find the final summed-up descriptor offset of the element
+    static uint32_t getDescOffsetFromPath(const ReflectionPath* pPath, uint32_t arraySize, SlangParameterCategory category)
+    {
+#ifndef FALCOR_VK
+        return 0;
+#else
+        uint32_t offset = 0;
+        bool first = true;
+        for (auto pp = pPath; pp; pp = pp->pParent)
+        {
+            if ((pp->pTypeLayout) && (pp->pTypeLayout->getKind() == TypeReflection::Kind::Array))
+            {
+                offset += (uint32_t)pp->childIndex * arraySize;
+                arraySize *= (uint32_t)pp->pTypeLayout->getElementCount();
+            }
+        }
+        return offset;
+#endif
+    }
+
+    static size_t getRegisterIndexFromPath(const ReflectionPath* pPath, SlangParameterCategory category)
+    {
+        uint32_t offset = 0;
+        for (auto pp = pPath; pp; pp = pp->pParent)
+        {
+            if (pp->pVar)
+            {
+                const auto& h = pp->pVar->getName();
+                offset += (uint32_t)pp->pVar->getOffset(category);
+                continue;
+            }
+            else if (pp->pTypeLayout)
+            {
+                switch (pp->pTypeLayout->getKind())
+                {
+                case TypeReflection::Kind::Array:
+                    offset += (uint32_t)pp->pTypeLayout->getElementStride(category) * pp->childIndex;
+                    continue;
+
+                case TypeReflection::Kind::Struct:
+                    offset += (uint32_t)pp->pTypeLayout->getFieldByIndex(int(pp->childIndex))->getOffset(category);
+                    continue;
+
+                default:
+                    break;
+                }
+            }
+
+            logError("internal error: invalid reflection path");
+            return 0;
+        }
+        return offset;
+    }
+
+    static size_t getUniformOffset(const ReflectionPath* pPath)
+    {
+        return getRegisterIndexFromPath(pPath, SLANG_PARAMETER_CATEGORY_UNIFORM);
+    }
+
+    static uint32_t getRegisterSpaceFromPath(const ReflectionPath* pPath, SlangParameterCategory category)
+    {
+        return (uint32_t)pPath->pVar->getBindingSpace(category);
```

Comment:
Eventually this needs to follow the same pattern as `getRegisterIndexFromPath` and add up the contributions that all parent nodes make to the register space. We can cross that bridge when we come to it, though.